Docker container

aprokop opened the pull request:
- Updating clang to 8.0 will require updating clang-format
- Not tested yet as there is no build system in place
- Unclear what version of OpenMPI to use
  Would like to match to Spectrum MPI installed on Summit

dalg24 commented on docker/.gitignore:
```diff
@@ -0,0 +1,3 @@
+docker-compose.override.yml
+Dockerfile_*_generated
```
```suggestion
```

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+        git \
+        wget \
+        vim \
+        emacs \
```
```suggestion
```

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+RUN apt-get update && apt-get install -y \
+        build-essential \
+        bc \
+        gfortran \
```
```suggestion
```

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+ARG BASE=ubuntu:16.04
```
```suggestion
ARG BASE=ubuntu:18.04
```

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+    add-apt-repository -y ppa:ubuntu-toolchain-r/test && \
+    apt-get update && apt-get install -y \
+        gcc-7 \
+        g++-7 && \
```
gcc-7 is the default on Ubuntu18.04

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+ENV MPI_DIR=/opt/openmpi/3.1
+# Put OPENMPI_DIR at the end of the path so that /ust/local/bin/mpiexec will
+# overwrite it
+ENV PATH=$PATH:${MPI_DIR}/bin
```
Prepend it

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+    tar -xf ${OPENMPI_ARCHIVE} -C ${OPENMPI_SOURCE_DIR} --strip-components=1 && \
+    mkdir -p ${OPENMPI_BUILD_DIR} && \
+    cd ${OPENMPI_BUILD_DIR} && \
+    ${OPENMPI_SOURCE_DIR}/configure --prefix=${OPENMPI_INSTALL_DIR} && \
```
Don't we need a flag for CUDA-aware MPI?

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+
+# install Kokkos Profiling and Debugging Tools
+# sh does not support arrays so we need to use bash
+# Note: the commit hash provided below has been tagged as Version 2.5.00 on GitHub
```
Why did you not match the Kokkos version?

dalg24 commented on docker/Dockerfile_stack:
```diff
@@ -0,0 +1,219 @@
+RUN echo '#!/usr/bin/env bash' > /usr/local/bin/mpiexec && \
+    echo '${MPI_DIR}/bin/mpiexec --allow-run-as-root "$@"' >> /usr/local/bin/mpiexec && \
+    chmod +x /usr/local/bin/mpiexec && \
+    mpiexec --version
```
Remove that block

aprokop commented:
Chatted with some OLCF folks:
- There is no easy way to find with version of OpenMPI IBM's Spetrum-MPI matches
- From some chats with IBM right now it seems to be version 3
- There may be some plans of IBM to move to 4 later in the year
- But there is not guarantee that increment in major version of Spectrum-MPI mean incrementing major version of OpenMPI

aprokop commented:
@dalg24 Addressed your comments. So far, tested non-cuda container only. DistrubutedSearchTree test segfaults, have not investigated why yet. Things to do:
- Decide on how many containers we want to build
  DTK builds 3, Cabana 1, I think at least two (cuda and non-cuda)
- How to organize files
  Where to put build script, how to name directories, etc.

dalg24 commented:
Try to revert cf431e2d084b8cd9318136f228f3ddd2652d8e5e

dalg24 commented:
retest this please

dalg24 commented:
retest this please

dalg24 commented:
retest this please

dalg24 commented:
retest this please

dalg24 commented:
retest this please

dalg24 commented:
NOT ready.  Just trying to figure CI out...

dalg24 commented:
test this please

dalg24 commented:
retest this please

dalg24 commented:
retest this please

dalg24 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,124 @@
+    OPENMPI_SHA1=b3c60e2bdd5a8a8e758fd741f9a5bebb84da5e81 && \
+    OPENMPI_URL=https://download.open-mpi.org/release/open-mpi/v${OPENMPI_VERSION_SHORT}/openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    OPENMPI_ARCHIVE=openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    [ ! -z "${CUDA_VERSION}" ] && CUDA_OPTIONS=--with-cuda || true && \
```
Defined in `nvidia/cuda`

dalg24 commented on docker/jenkins/build.sh:
```diff
@@ -0,0 +1,32 @@
+
+make -j${NPROCS}
+
+ctest --no-compress-output -T Test
```
Yes we have not set the number of procs/threads required by each test.

dalg24 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,124 @@
+    [ ! -z "${CUDA_VERSION}" ] && CUDA_OPTIONS="--with-cuda --with-cuda-options=enable_lambda,enable_uvm" || true && \
+    ../generate_makefile.bash --prefix=${KOKKOS_DIR} \
+       --with-serial --with-openmp ${CUDA_OPTIONS} \
+       --arch=SNB,Volta70 \
```
Will make it a build argument in a later PR

dalg24 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,124 @@
+ENV LLVM_DIR=/opt/llvm
+RUN LLVM_VERSION=7.0.1 && \
+    LLVM_URL=http://releases.llvm.org/${LLVM_VERSION}/clang+llvm-${LLVM_VERSION}-x86_64-linux-gnu-ubuntu-18.04.tar.xz && \
+    LLVM_ARCHIVE=llvm-${LLVM_VERSION}.tar.xz && \
```
Now verifying the signature

dalg24 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,124 @@
+    OPENMPI_URL=https://download.open-mpi.org/release/open-mpi/v${OPENMPI_VERSION_SHORT}/openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    OPENMPI_ARCHIVE=openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    [ ! -z "${CUDA_VERSION}" ] && CUDA_OPTIONS=--with-cuda || true && \
+    export SCRATCH_DIR=${PWD}/scratch && mkdir -p ${SCRATCH_DIR} && cd ${SCRATCH_DIR} && \
```
Fixed

dalg24 commented:
@aprokop I have stripped the docker-compose for the dev container
I will add it along with a configure script in another PR
